lib/backend/k8s/resources/node.go

```diff
@@ -300,6 +303,76 @@ func K8sNodeToCalico(k8sNode *kapiv1.Node, usePodCIDR bool) (*model.KVPair, erro
 	}, nil
 }
 
+func listAllAddresses(k8sNode *kapiv1.Node, calicoNode *apiv3.Node) []apiv3.NodeAddress {
+	var addrs []apiv3.NodeAddress
+
+	if bgp := calicoNode.Spec.BGP; bgp != nil {
+		if addr := bgp.IPv4Address; addr != "" {
+			addrs = append(addrs, apiv3.NodeAddress{Address: addr, Type: apiv3.NodeAddressIP})
+		}
+		if addr := bgp.IPv6Address; addr != "" {
+			addrs = append(addrs, apiv3.NodeAddress{Address: addr, Type: apiv3.NodeAddressIP})
+		}
+		if addr := bgp.IPv4IPIPTunnelAddr; addr != "" {
+			addrs = append(addrs, apiv3.NodeAddress{Address: addr, Type: apiv3.NodeAddressIPIPTunnelIP})
+		}
+	}
+
+	if addr := calicoNode.Spec.IPv4VXLANTunnelAddr; addr != "" {
+		addrs = append(addrs, apiv3.NodeAddress{Address: addr, Type: apiv3.NodeAddressVxlanTunnelIP})
+	}
+	if addr := calicoNode.Spec.VXLANTunnelMACAddr; addr != "" {
+		addrs = append(addrs, apiv3.NodeAddress{Address: addr, Type: apiv3.NodeAddressVxlanMAC})
+	}
+
+	if wg := calicoNode.Spec.Wireguard; wg != nil {
+		addrs = append(addrs,
+			apiv3.NodeAddress{
+				Address: wg.InterfaceIPv4Address,
+				Type:    apiv3.NodeAddressWireguardTunnelIP,
+			},
+		)
+	}
+
+	for _, kaddr := range k8sNode.Status.Addresses {
+		var addr apiv3.NodeAddress
+		switch kaddr.Type {
+		case kapiv1.NodeInternalIP:
+			addr = apiv3.NodeAddress{
+				Type:    apiv3.NodeAddressK8sInternalIP,
+				Address: kaddr.Address,
+			}
+		case kapiv1.NodeExternalIP:
+			addr = apiv3.NodeAddress{
+				Type:    apiv3.NodeAddressK8sExternalIP,
+				Address: kaddr.Address,
+			}
+		default:
+			continue
+		}
+
+		// k8s InternalIP and possibly ExternalIP are likely already in the list
+		// from the BPG spec IPv4 or IPv6 fields. Find them and upgrade them to
+		// the k8s type.
```

Comment:
Think I'd leave our this step; the client can dedupe if it matches their use case but, in general, I prefer to have the raw data in the datamodel and helper functions to massage it on read if that's useful.  (It's difficult to rejig the datamodel later if we decide that we've lost some information.)

Reply:
I was considering that, and than I opted for deduping here since each user, instead of just looping over the list would need to check if it is already seen an address, not to use it 2x - which may make using this more difficult. My reasoning was that "upgrading" `NodeAddressIP` to `NodeAddressK8sXXXIP` only adds information. It is still an IP, but also known to K8s.

If we pass the an address 2x, should we also make the `NodeAddressIP` to be more precise like `NodeAddressBGPIP?